### PR TITLE
Add internal solution for Maximum Subsequence (cf-888e)

### PR DESCRIPTION
**Pull request by ChuangSheep**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/cf-888E.mdx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+## Explanation
+
+For this problem, we are given $n \leq 35$ numbers and have to find a combination such that the sum of all elements in this combination modulo $m$ is maximized. A naive approach, i.e. trying out all $2^{35}$ possible combinations, is not feasible. However, we can use the [Meet-In-The-Middle Technique](https://usaco.guide/gold/meet-in-the-middle) to reduce the time complexity to $2^{\frac{n}{2}}$, which would be enough for our time bound. 
+
+We first split the given array into two halves that are equally long. Then, we generate all combinations of numbers in the first half by using a bit mask and save the results modulo $m$ in a sorted set. For the second half, let's also generate all combinations for them. For each of these combinations, the maximal sum is the sum of all elements in this combination, $current\_sum$, plus the largest sum of a combination of the first half, $left\_sum$, which is less than $m - current\_sum$. It is never optimal to choose a value larger than $m - current\_sum$ since otherwise the result would be $left\_sum + current\_sum - m$, which is strictly less than $current\_sum$.
```
Comment (SansPapyrus683) [inline solutions/cf-888E.mdx:14]: > We first split the given array into two halves that are equally long.

what if the array length is odd?

File: solutions/cf-888E.mdx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+using namespace std;
+
+/**
+ * Calculate the sum of all elements in arr, represented by the binary mask, and
+ * take modulo mod.
+ */
+int unmask(int mask, const vector<int> &arr, int mod) {
+	int current_sum = 0;
+	for (int bit = 0; bit < arr.size(); bit++) {
+		if (mask >> bit & 1 == 1) {
```
Comment (SansPapyrus683) [inline solutions/cf-888E.mdx:34]: you sure the order of operators works for this? (just making sure)

File: solutions/cf-888E.mdx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	// the best value from all combinations of left_arr
+	int best = *left_sums.rbegin();
+	for (int mask = 0; mask < 1 << right_arr.size(); mask++) {
+		int current_sum = unmask(mask, right_arr, m);
+		/*
+		 * a possible new maximum value is the sum of current_sum and the largest
+		 * value below m - current_sum from the combinations in left_arr
+		 */
+		best =
+			max(best, *prev(left_sums.lower_bound(m - current_sum)) + current_sum);
```
Comment (SansPapyrus683) [inline solutions/cf-888E.mdx:71]: sus

Comment (SansPapyrus683) [reply solutions/cf-888E.mdx:71]: i usually format long stuff like this:
```cpp
x = max(
	a,
	b
);

File: solutions/cf-888E.mdx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	cout << best;
```
Comment (SansPapyrus683) [inline solutions/cf-888E.mdx:74]: cout << best << endl;

File: solutions/cf-888E.mdx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+}
+```
+</CPPSection>
+
+<JavaSection>
+```java
+import java.io.*;
+import java.util.*;
+
+public class MaximumSubsequence {
```
Comment (SansPapyrus683) [inline solutions/cf-888E.mdx:84]:
```suggestion
public class MaxSubseq {
```

File: solutions/cf-888E.mdx [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		StringTokenizer st = new StringTokenizer(br.readLine());
+		int n = Integer.parseInt(st.nextToken());
+		int m = Integer.parseInt(st.nextToken());
+
+		int[] arr = Arrays.stream(br.readLine().split(" "))
+				.mapToInt((i) -> Integer.parseInt(i))
```
Comment (SansPapyrus683) [inline solutions/cf-888E.mdx:92]:
```suggestion
				.mapToInt(Integer::parseInt)
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
nice!

File: solutions/cf-888E.mdx (hunk 2 above)
Comment (ChuangSheep) [reply solutions/cf-888E.mdx:34]: lol it works (and ac), but not the way I wanted. It does `((mask >> bit) & (1 == 1))` which gives the right answer, but ima gonna add parentheses to make it work "properly"

File: solutions/cf-888E.mdx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+	// stores the sums of all combinations from the left_arr modulo m
+	set<int> left_sums;
+	for (int mask = 0; mask < (1 << left_arr.size()); mask++) {
+		left_sums.insert(unmask(mask, left_arr, m));
+	}
+
+	// the best value from all combinations of left_arr
+	int best = *left_sums.rbegin();
+	for (int mask = 0; mask < 1 << right_arr.size(); mask++) {
```
Comment (SansPapyrus683) [inline solutions/cf-888E.mdx:64]: parens here for the bitwise too?

**Review by SansPapyrus683 — APPROVED** [review]
naisu

**Review by ryanchou-dev — APPROVED** [review]
perfect

**Pull request merged**